### PR TITLE
Add edit button to discovered layout dialog

### PR DESCRIPTION
**Pull request by mkondratek**
- revamp DiscoverTableDialog from `.java` to `.kt`
- fix listing commits (someone (me) accidentally turned off commits within demo dialog while turning it off within discovered dialog :xd:)
- Rename button `Save Discovered Layout` to `Save` 
- add `Edit` Button & action
- properly handle not found machete file (actually I see no real-life possibility to achieve that but anyway)

image

### CODE REVIEW

File: frontendResourceBundles/src/main/resources/GitMacheteBundle.properties [hunk 1]
```diff
@@ -263,6 +263,8 @@ action.GitMachete.OpenMacheteFileAction.description=Open machete file
 
 action.GitMachete.OpenMacheteFileAction.notification.fail.cannot-open=Failed to open machete file
 action.GitMachete.OpenMacheteFileAction.notification.fail.same-name-dir-exists=Cannot create file 'machete': Directory with the same name exists
+action.GitMachete.OpenMacheteFileAction.notification.fail.machete-file-not-found=Machete file has not been found
+action.GitMachete.OpenMacheteFileAction.notification.message.machete-file-not-found=Could not found machete file under specified root ({0})
```
Comment (PawelLipski) [inline frontendResourceBundles/src/main/resources/GitMacheteBundle.properties:267]: Nit:
```suggestion
action.GitMachete.OpenMacheteFileAction.notification.message.machete-file-not-found=Could not found machete file under the specified root ({0})
```

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/toolbar/OpenMacheteFileAction.java [hunk 2]
```diff
@@ -55,12 +56,16 @@ public void actionPerformed(AnActionEvent anActionEvent) {
     getGraphTable(anActionEvent).queueRepositoryUpdateAndModelRefresh();
 
     if (macheteFile.isEmpty()) {
-      log().warn("Skipping the action because machete file is undefined");
+      VcsNotifier.getInstance(project).notifyError(
+          /* title */ getString("action.GitMachete.OpenMacheteFileAction.notification.fail.machete-file-not-found"),
```
Comment (PawelLipski) [inline frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/toolbar/OpenMacheteFileAction.java:60]: Hmmm `.fail.` vs `.message.`? Maybe `.title.` or `.fail.{title,notification}.` instead?

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+                  cancelButtonVisible,
+                  windowTitle,
+                  okButtonText)
+            }
+
+    fun ofDemoRepository() =
+        RuntimeBinding.instantiateSoleImplementingClass(ISimpleGraphTableProvider::class.java)
+            .deriveDemoInstance()
+            .let {
+          GraphTableDialog(
```
Comment (PawelLipski) [inline frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt:66]: Whoops weird formatting (?)

File: frontendUiApi/src/main/java/com/virtuslab/gitmachete/frontend/ui/api/table/ISimpleGraphTableProvider.java [hunk 4]
```diff
@@ -6,7 +6,7 @@
 
 public interface ISimpleGraphTableProvider {
   @UIEffect
-  BaseGraphTable deriveInstance(IGitMacheteRepositorySnapshot macheteRepositorySnapshot);
+  BaseGraphTable deriveInstance(IGitMacheteRepositorySnapshot macheteRepositorySnapshot, boolean toggleListingCommits);
```
Comment (PawelLipski) [inline frontendUiApi/src/main/java/com/virtuslab/gitmachete/frontend/ui/api/table/ISimpleGraphTableProvider.java:9]: Hmm why this change is on this PR? :thinking:

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/graphtable/DiscoverAction.java [hunk 5]
```diff
@@ -52,22 +55,40 @@ public void actionPerformed(AnActionEvent anActionEvent) {
...
+    }, NON_MODAL);
+  }
+
   private void saveDiscoveredLayout(IGitMacheteRepositorySnapshot repositorySnapshot, Path macheteFilePath, Project project,
       BaseEnhancedGraphTable baseEnhancedGraphTable, IBranchLayoutWriter branchLayoutWriter) {
     var branchLayout = repositorySnapshot.getBranchLayout().getOrNull();
     if (branchLayout == null) {
       VcsNotifier.getInstance(project).notifyError(
-          /* title */ getString("action.GitMachete.DiscoverAction.cant-discover-layout-error-title"), /* message */ "");
+          /* title */ getString("action.GitMachete.DiscoverAction.cant-discover-layout-error-title"),
```
Comment (PawelLipski) [inline frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/graphtable/DiscoverAction.java:90]: Yeah, pls make the usage of `-title`/`.title.` consistent in the notifications

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt [hunk 6]
```diff
@@ -0,0 +1,107 @@
+package com.virtuslab.gitmachete.frontend.actions.dialogs
```
Comment (PawelLipski) [inline frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt:1]: Given that there's quite a few non-trivial changes here... pls bump the version to v0.7.1 (and I'll move issue/PR now)

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+  override fun createActions() =
+      if (cancelButtonVisible)
+          arrayOf(getEditAction(), getOKAction(), cancelAction).filterNotNull().toTypedArray()
+      else arrayOf(getOKAction())
+
+  private fun getEditAction() =
+      if (editAction == null) null
+      else
+          object : AbstractAction("Edit") {
```
Comment (PawelLipski) [inline frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt:86]: Hmm actually that's `Save & Edit` action to be precise (it does the same as `Save` and additionally opens the editor)... not sure if we want to reflect this nuance in the UI

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+            override fun actionPerformed(e: ActionEvent?) {
+              doOKAction()
+              editAction.run()
+              close(OK_EXIT_CODE)
```
Comment (PawelLipski) [inline frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt:90]: Hmm but `doOKAction()` already calls `close(OK_EXIT_CODE)`?  :thinking:

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+  override fun createCenterPanel() =
+      JBUI.Panels.simplePanel(/* hgap */ 0, /* vgap */ 2).apply {
+        addToCenter(ScrollPaneFactory.createScrollPane(this@GraphTableDialog.table))
+        preferredSize = dimension
+      }
+
+  @Override
+  override fun doOKAction() {
+    if (getOKAction().isEnabled) {
```
Comment (PawelLipski) [inline frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt:102]: I don't get it, how can `OKAction` be NOT enabled when `doOKAction` is called? :thinking:

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    // @JvmOverloads annotation instructs the Kotlin compiler to generate overloads
+    // for this function that substitute default parameter values (dimension in our case).
+    @JvmOverloads
+    fun of(
+        gitMacheteRepositorySnapshot: IGitMacheteRepositorySnapshot,
+        windowTitle: String,
+        dimension: JBDimension = JBDimension(/* width */ 800, /* height */ 500),
+        emptyTableText: String?,
+        okAction: Consumer<IGitMacheteRepositorySnapshot>?,
+        editAction: Runnable?,
```
Comment (PawelLipski) [inline frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt:43]: There's (?) a silent assumption that `okAction` invokes `editAction`... :thinking:

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+        okButtonText: String,
+        cancelButtonVisible: Boolean
+    ) =
+        RuntimeBinding.instantiateSoleImplementingClass(ISimpleGraphTableProvider::class.java)
+            .deriveInstance(gitMacheteRepositorySnapshot, /* toggleListingCommits */ false)
+            .apply { setTextForEmptyTable(emptyTableText ?: "") }
+            .let {
+              GraphTableDialog(
+                  it,
```
Comment (PawelLipski) [inline frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt:52]: Maybe name `it` here for clarity

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+    ) : DialogWrapper(/* canBeParent */ false) {
+
+  init {
+    super.init()
+    title = windowTitle
+    setOKButtonText(okButtonText)
+  }
+
+  companion object {
+    // @JvmOverloads annotation instructs the Kotlin compiler to generate overloads
```
Comment (PawelLipski) [inline frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt:34]: :+1:

File: frontendUiApi/src/main/java/com/virtuslab/gitmachete/frontend/ui/api/table/ISimpleGraphTableProvider.java (hunk 4 above)
Comment (mkondratek) [reply frontendUiApi/src/main/java/com/virtuslab/gitmachete/frontend/ui/api/table/ISimpleGraphTableProvider.java:9]: unrel, just fixing help dialog

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt (hunk 3 above)
Comment (mkondratek) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt:66]: here looks better
image

a bit weird but enforced by spotless :confused:

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt (hunk 9 above)
Comment (mkondratek) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt]: right

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt (hunk 7 above)
Comment (mkondratek) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt]: right, let it be short but precise

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt (hunk 8 above)
Comment (mkondratek) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt]: good point

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt (hunk 10 above)
Comment (mkondratek) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt:43]: y? I do not get your doubts :thinking:

File: frontendUiApi/src/main/java/com/virtuslab/gitmachete/frontend/ui/api/table/ISimpleGraphTableProvider.java (hunk 4 above)
Comment (mkondratek) [reply frontendUiApi/src/main/java/com/virtuslab/gitmachete/frontend/ui/api/table/ISimpleGraphTableProvider.java:9]: as this one is gonna be 0.7.1 we may want to ship it on the separate pr for 0.7.0

Comment (micpiotrowski) [reply frontendUiApi/src/main/java/com/virtuslab/gitmachete/frontend/ui/api/table/ISimpleGraphTableProvider.java:9]: Ok, I can

Comment (micpiotrowski) [reply frontendUiApi/src/main/java/com/virtuslab/gitmachete/frontend/ui/api/table/ISimpleGraphTableProvider.java:9]: Done

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt (hunk 10 above)
Comment (mkondratek) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt:43]: that's right. We could build both build "actions" within the `DiscoverAction` and gain the isolation here. Is it your point?

File: frontendResourceBundles/src/main/resources/GitMacheteBundle.properties [hunk 13]
```diff
@@ -319,5 +321,5 @@ action.GitMachete.ToggleListingCommitsAction.description.disabled.no-branches=To
 action.GitMachete.OpenMacheteTabAction.text=Open Git Machete Tab
 action.GitMachete.OpenMacheteTabAction.description=Open Git Machete Tab
 
-action.GitMachete.OpenMacheteTabAction.notification.fail.title=Could not open Git Machete tab
-action.GitMachete.OpenMacheteTabAction.notification.fail.message.no-git=Ensure that Git VCS is enabled in the project
+action.GitMachete.OpenMacheteTabAction.notification.title.could-not-open-tab=Could not open Git Machete tab
```
[the remark this answers is not in the document]
Comment (mkondratek) [reply frontendResourceBundles/src/main/resources/GitMacheteBundle.properties:324]: ok, so what's about this rule:
> We use `cannot` if "it" (whatever IT is) can be never performed but `could not` if "it" cannot due to some exceptional circumstances.

Hence:
- `Cannot create file 'machete': Directory with the same name exists` b/c it is never possible while the directory exists
- `Could not open Git Machete tab` b/c we are not able to specify why it was not opened; it could be missing tool window or missing content(tab) but we neither want to expose the technical terms nor we know the actual reason why the tool window or the content tab is null; most likely it's disabled Git Vcs

[the remark this answers is not in the document]
Comment (PawelLipski) [reply frontendResourceBundles/src/main/resources/GitMacheteBundle.properties:324]: Ok! Pls apply & document in .properties file or in CONTRIBUTING.md

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt (hunk 10 above)
Comment (PawelLipski) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt:43]: Yes, would be nice

**Comment by PawelLipski** [comment]
**On hold**: #545 can make this PR obsolote

**Comment by micpiotrowski** [comment]
This work still have a sense coz discover dialog survive my PR #550 - **on-hold removed** ;)

File: frontendResourceBundles/src/main/resources/GitMacheteBundle.properties [hunk 14]
```diff
@@ -3,6 +3,17 @@
 #   which are rendered by MessageFormat.format as single single quotes)
 # Notification messages - <b></b> tags allowed (WE use <b></b> for indication)
 # Dialogs messages (aka content) - HTML wrapping required for tags usage (WE use <b></b> for indication)
+#
+# Could not / Cannot usage
+# We use cannot if "it" (whatever IT is) can be never performed
```
Comment (PawelLipski) [inline frontendResourceBundles/src/main/resources/GitMacheteBundle.properties:8]:
```suggestion
# We use `cannot` if "it" (whatever IT is) can be never performed
```

File: frontendResourceBundles/src/main/resources/GitMacheteBundle.properties [hunk 15]
```diff
@@ -3,6 +3,17 @@
...
+# but could not if "it" cannot due to some exceptional circumstances.
```
Comment (PawelLipski) [inline frontendResourceBundles/src/main/resources/GitMacheteBundle.properties:9]: ?
```suggestion
# but `could not` if "it" could not have been performed due to some exceptional circumstances.
```

File: frontendResourceBundles/src/main/resources/GitMacheteBundle.properties [hunk 16]
```diff
@@ -3,6 +3,17 @@
...
+#Hence:
```
Comment (PawelLipski) [inline frontendResourceBundles/src/main/resources/GitMacheteBundle.properties:10]:
```suggestion
# Hence:
```

File: frontendResourceBundles/src/main/resources/GitMacheteBundle.properties [hunk 17]
```diff
@@ -3,6 +3,17 @@
...
+# - "Cannot create file 'machete': Directory with the same name exists"
+# because it is never possible while the directory exists
+# - "Could not open Git Machete tab"
+# because we are not able to specify why it was not opened; it could be missing tool window or missing content(tab)
+# but we neither want to expose the technical terms nor we know the actual reason
```
Comment (PawelLipski) [inline frontendResourceBundles/src/main/resources/GitMacheteBundle.properties:15]:
```suggestion
# but we neither want to expose the technical details nor we know the actual reason
```

**Review by PawelLipski — APPROVED** [review]
Minor comments, approving

**Review by PawelLipski — APPROVED** [review]
